File: frontend/src/types/itinerary.ts
import type { Event } from './event'

export interface ItineraryItem {
  id: number
  event_id: number | null
  day_index: number
  custom_note: string | null
  event: Event | null

  custom_title: string | null
  custom_location: string | null
  custom_start_time: string | null
  custom_end_time: string | null
  custom_notes: string | null
}

export interface Itinerary {
  id: number
  name: string
  start_date: string
  end_date: string
  created_at: string
  items: ItineraryItem[]
}

export interface ItineraryCreate {
  name: string
  start_date: string
  end_date: string
}

export interface ItineraryItemCreate {
  event_id?: number
  day_index: number
  custom_note?: string

  custom_title?: string
  custom_location?: string
  custom_start_time?: string
  custom_end_time?: string
  custom_notes?: string
}

File: backend/app/models/itinerary.py
from sqlalchemy import Column, Integer, String, Text, DateTime, ForeignKey
from sqlalchemy.orm import relationship
from app.database import Base
from datetime import datetime, timezone


class Itinerary(Base):
    __tablename__ = "itineraries"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    name = Column(String, nullable=False)
    start_date = Column(DateTime, nullable=False)
    end_date = Column(DateTime, nullable=False)
    created_at = Column(DateTime, default=lambda: datetime.now(timezone.utc))

    items = relationship(
        "ItineraryItem", back_populates="itinerary", cascade="all, delete-orphan")
    user = relationship("User", back_populates="itineraries")


class ItineraryItem(Base):
    __tablename__ = "itinerary_items"

    id = Column(Integer, primary_key=True, index=True)
    itinerary_id = Column(Integer, ForeignKey("itineraries.id"), nullable=False)
    event_id = Column(Integer, ForeignKey("events.id"), nullable=True)
    day_index = Column(Integer, nullable=False)
    custom_note = Column(Text, nullable=True)

    # Custom (user-defined) event fields
    custom_title = Column(String, nullable=True)
    custom_location = Column(String, nullable=True)
    custom_start_time = Column(DateTime, nullable=True)
    custom_end_time = Column(DateTime, nullable=True)
    custom_notes = Column(Text, nullable=True)

    itinerary = relationship("Itinerary", back_populates="items")
    event = relationship("Event", foreign_keys=[event_id])
